pkg/reconciler/v1alpha1/pipelinerun/resources/apply.go

```diff
@@ -0,0 +1,65 @@
+/*
+ Copyright 2019 Knative Authors LLC
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// ApplyParameters applies the params from a PipelineRun.Params to a PipelineSpec.
+func ApplyParameters(p *v1alpha1.Pipeline, pr *v1alpha1.PipelineRun, defaults ...v1alpha1.PipelineParam) *v1alpha1.Pipeline {
+	// This assumes that the PipelineRun inputs have been validated against what the Pipeline requests.
+	replacements := map[string]string{}
+	// Set all the default replacements
+	for _, p := range defaults {
+		if p.Default != "" {
+			replacements[fmt.Sprintf("params.%s", p.Name)] = p.Default
+		}
+	}
+	// Set and overwrite params with the ones from the PipelineRun
+	for _, p := range pr.Spec.Params {
+		replacements[fmt.Sprintf("params.%s", p.Name)] = p.Value
+	}
+
+	return ApplyReplacements(p, replacements)
+}
+
+// ApplyReplacements replaces placeholders for declared parameters with the specified replacements.
+func ApplyReplacements(p *v1alpha1.Pipeline, replacements map[string]string) *v1alpha1.Pipeline {
+	p = p.DeepCopy()
+
+	applyReplacements := func(in string) string {
```

Comment:
I know this feels a bit dumb but we should extract this function into a package so that `taskrun/resources/apply.go` and `pipelinerun/resources/apply.go` can share it (and if we make the syntax evolve, it evolves for both). wdyt ?

Reply:
Sounds good to me - where do you think that should go?

Reply:
I like the idea of extracting some of the common logic into a package.  This is just a suggestion on where i think it should be present feel free to propose if you think some other solution is better. 
It could be a package at level of pipelinerun and taskrun. We expect the package to be used by both of these sub packages.

Reply:
Ok, I'll give that a try. Thanks!